cmd/tcplb/server.go

```
@@ -0,0 +1,220 @@
+	MaxConnectionsPerClient int64
+}
+
+func (c *Config) Validate() error {
+	if len(c.Upstreams) == 0 {
+		return errors.New("server must be configured with 1 or more upstreams")
+	}
+	return nil
+}
+
+func makeClientReserverFromConfig(cfg *Config) (forwarder.ClientReserver, error) {
+	var reserver forwarder.ClientReserver
+	if cfg.MaxConnectionsPerClient > 0 {
+		reserver = limiter.NewUniformlyBoundedClientReserver(cfg.MaxConnectionsPerClient)
+	} else {
+		reserver = limiter.UnboundedClientReserver{}
+	}
+	return reserver, nil
+}
+
+func makeAuthorizerFromConfig(cfg *Config) (forwarder.Authorizer, error) {
+	// TODO FIXME begin placeholder demo authorization config
+	urGroup := authz.Group{Key: "ur"}
+	urUpstreamGroup := authz.UpstreamGroup{Key: "ur"}
+	authzCfg := authz.Config{
+		GroupsByClientID: map[core.ClientID][]authz.Group{
+			anonymousTestClientID: {urGroup},
+		},
+		UpstreamGroupsByGroup: map[authz.Group][]authz.UpstreamGroup{
+			urGroup: {urUpstreamGroup},
+		},
+		UpstreamsByUpstreamGroup: map[authz.UpstreamGroup]core.UpstreamSet{
+			urUpstreamGroup: core.NewUpstreamSet(cfg.Upstreams...),
+		},
+	}
+	// TODO FIXME end placeholder demo authorization config
+	return authz.NewStaticAuthorizer(authzCfg), nil
+}
+
+// IdiotDialer attempts to dial an arbitrary candidate and gives up if that fails.
+// This is a placeholder implementation with various issues:
+// - no timeout
+// - it doesn't attempt to balance load
+// - it doesn't try alternative upstreams if one attempt fails
+// - it doesn't learn anything
+type IdiotDialer struct{}
+
+func (d IdiotDialer) DialBestUpstream(ctx context.Context, candidates core.UpstreamSet) (core.Upstream, forwarder.DuplexConn, error) {
+	for c := range candidates {
+		conn, err := net.Dial(c.Network, c.Address)
+		if err != nil {
+			return core.Upstream{}, nil, err
+		}
+		switch upstreamConn := conn.(type) {
+		case *net.TCPConn:
+			return c, upstreamConn, nil
+		default:
+			_ = conn.Close()
+			break
+		}
```

Maybe this is worthy of a log message? Not that it should ever happen, but silently closed connections could be surprising.

---

good idea. the plan is to replace this implementation, but adding an error message costs 5 minutes now, trying to debug something failing silently there would cost > 5 minutes.